proposed/0023-immutable-data-type-naming/0023-immutable-data-type-naming.md

```diff
@@ -0,0 +1,113 @@
+
+## Secondary Benefits
+
+The proposed solution has a useful side-effect: for a given name type, both of the other names can be deduced.  This cannot be done for hashing; given a name derived from a hash digest, we cannot calculate the original name, i.e. the hash cannot be reversed.
+
+This is useful since a Vault which is managing a Backup or Sacrificial copy of a chunk would now be able to contact the Normal managers without requiring an actual copy of the chunk.
+
+A further benefit is that XOR is a faster operation than SHA512 hash; in benchmarks, using the proposed solution was between 15% and 20% faster than using the existing implementation.
+
+# Detailed Design
+
+The proposed change would define the following constants and function:
+
+```rust
+const NORMAL_TO_BACKUP: [u8; XOR_NAME_LEN] =
+    [128, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
+     0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
+     0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0];
+const NORMAL_TO_SACRIFICIAL: [u8; XOR_NAME_LEN] = [255; XOR_NAME_LEN];
+const BACKUP_TO_SACRIFICIAL: [u8; XOR_NAME_LEN] =
+    [127, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255,
+     255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255,
+     255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255,
+     255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255,
+     255, 255, 255, 255];
+
+fn xor(lhs: &[u8; XOR_NAME_LEN], mut rhs: [u8; XOR_NAME_LEN]) -> XorName {
+    for i in 0..XOR_NAME_LEN {
+        rhs[i] ^= lhs[i];
+    }
+    XorName(rhs)
+}
+```
+
+These would be used in `ImmutableData`'s implementation as follows:
+
+```rust
+impl ImmutableData {
+    ...
+    pub fn name(&self) -> XorName {
+        let digest = sha512::hash(&self.value);
+        match self.type_tag {
+            ImmutableDataType::Normal => XorName(digest.0),
+            ImmutableDataType::Backup => xor(&digest.0, NORMAL_TO_BACKUP),
+            ImmutableDataType::Sacrificial => xor(&digest.0, NORMAL_TO_SACRIFICIAL),
+        }
+    }
+}
+```
+
+It would also allow us to define the following functions to convert between name types:
+
+```rust
+pub fn normal_to_backup(name: &XorName) -> XorName {
+    xor(&name.0, NORMAL_TO_BACKUP)
+}
+
+pub fn backup_to_normal(name: &XorName) -> XorName {
+    xor(&name.0, NORMAL_TO_BACKUP)
+}
+
+pub fn normal_to_sacrificial(name: &XorName) -> XorName {
+    xor(&name.0, NORMAL_TO_SACRIFICIAL)
+}
+
+pub fn sacrificial_to_normal(name: &XorName) -> XorName {
+    xor(&name.0, NORMAL_TO_SACRIFICIAL)
+}
+
+pub fn backup_to_sacrificial(name: &XorName) -> XorName {
+    xor(&name.0, BACKUP_TO_SACRIFICIAL)
+}
+
+pub fn sacrificial_to_backup(name: &XorName) -> XorName {
+    xor(&name.0, BACKUP_TO_SACRIFICIAL)
+}
```

Comment:
This looks good and will prove much more efficient in code/speed and ability to provide a 2 way function as we desire.